kyo-core/shared/src/main/scala/kyo/scheduler/IOPromise.scala

```diff
@@ -1,151 +1,181 @@
 package kyo.scheduler
 
 import IOPromise.*
+import java.lang.invoke.MethodHandles
+import java.lang.invoke.VarHandle
+import java.util.concurrent.Executors
 import java.util.concurrent.atomic.AtomicReference
 import java.util.concurrent.locks.LockSupport
 import kyo.*
+import kyo.Result.Panic
+import kyo.kernel.Platform
+import kyo.kernel.Safepoint
+import kyo.scheduler.Scheduler
 import scala.annotation.tailrec
 import scala.util.control.NonFatal
+import scala.util.control.NoStackTrace
 
-private[kyo] class IOPromise[T](state: State[T])
-    extends AtomicReference(state):
+private[kyo] class IOPromise[E, A](init: State[E, A]) extends Safepoint.Interceptor:
+
+    @volatile private var state: State[E, A] = init
```

Comment:
I've changed the code to use `state` via `VarHandle` so it's possible to extend from the `Safepoint.Interceptor` class and avoid interface dispatches. Regular `IOPromise`s have the super class but as a no-op and `IOTask` overrides the no-op impls.